### ISSUE TITLE
Allow using fingerprint if it's registered

### ISSUE THREAD

**Issue by cassidyjames** [elementary/pantheon-agent-polkit#33]
I'm not sure how all of the PAM stuff works, but if a fingerprint is registered for the user, we should prefer that for authentication flows. It could look something like this:

![image](https://user-images.githubusercontent.com/611168/44890773-22e1d780-ac99-11e8-9e52-13563e4a238e.png)

"Use Password" would show the current password dialog instead.

**Comment by cassidyjames** [comment]
Sample dialog code from my mockup is at https://github.com/cassidyjames/agents/blob/master/src/BiometricDialog.vala

**Comment by arshubham** [comment]
This looks great as of today. But maybe the dialog mockup also take care of other authentication methods which elementary may support in the future like Pins, Hardware Keys, IR Cameras (If somebody brings Windows Hello support to linux). Just a thought to be future ready. Maybe it can have a dropdown for alternative authentication methods.

**Comment by cassidyjames** [comment]
@arshubham The code in my sample dialog is written in a way that could be easily expanded to multiple authentication types. An icon, authentication summary, and detailed description are passed in and used.

```vala
private const string BIO_AUTH_ICON = "fingerprint-symbolic";
private const string BIO_AUTH_SUMMARY = _("Touch");
private const string BIO_AUTH_DETAIL = _("Touch the fingerprint reader");
```

You can imagine this being expanded for a Yubikey:

```vala
private const string BIO_AUTH_ICON = "drive-removable-media-usb";
private const string BIO_AUTH_SUMMARY = _("Insert Key");
private const string BIO_AUTH_DETAIL = _("Insert your authenticator key");
```

which would give you something like:

![screenshot from 2018-08-31 10 52 05](https://user-images.githubusercontent.com/611168/44925537-ebf8d980-ad0b-11e8-8055-13266630129f.png)

**Comment by cassidyjames** [comment]
For the case of multiple possible authentication methods, I'm not sure what we'd do. That's definitely something to think about.

**Comment by arshubham** [comment]
> For the case of multiple possible authentication methods, I'm not sure what we'd do. That's definitely something to think about.
This was my query about. A device can have a muliple biometric authentication methods. I just wanted to point out that the dialog should me made in a future proof way preserving the choice of user. Great work by the way! :clap:

**Comment by danrabbit** [comment]
I think we probably need to come up with something that is a bit more agnostic to the form of authentication. It is much simpler if we can only ask for one at a time, but I think because we have a lot more options than on mobile that we're going to find ourselves in a situation where we're trying to create a hierarchy of asking the user which auth method they want to use in a series of dialogs which would get annoying fast if the method you want to use changes depending on external conditions (my hands are wet or my spouse has my password; I left my auth key at work; My phone is not near me so I can't use bluetooth auth). I think in an ideal world I can set up a few different types of auth and then use whichever is more convenient at the time right away.

I wonder if we can just append additional language to the current auth dialog. Forgive my shitty mockup, but something like this:

![bitmap](https://user-images.githubusercontent.com/7277719/44928096-a3412080-ad0b-11e8-860b-5bc77a5edc4b.png)

**Comment by cassidyjames** [comment]
@danrabbit yeah, that direction could work. I liked the idea of a super simple dialog for the simple case, but maybe this is a better route.

**Comment by danrabbit** [comment]
Yeah I think in this way that the dialog would only be as complicated as the users' auth setup. If they didn't configure any of these additional auth methods, they'd only see what we have right now. But if they added something else, we would explicitly let them know they can use it here without making them jump through any hoops to get to a particular auth method

**Comment by cassidyjames** [comment]
@danrabbit my code mockup has been updated at https://github.com/cassidyjames/agents/blob/master/src/BiometricDialog.vala

![screenshot from 2018-08-31 14 28 47](https://user-images.githubusercontent.com/611168/44934984-2f634000-ad2c-11e8-8057-be5dca8fd686.png)

**Comment by cassidyjames** [comment]
After a bit more iterating, we came to this:

![image 1](https://user-images.githubusercontent.com/611168/44937731-f761fa00-ad37-11e8-8682-14aef3fb6c4b.png)

1. Unique colors per authentication method. These would be used system-wide and could probably be defined in the system stylesheet for consistency.
2. A success state for "passive" authentication methods, like paired devices, already-connected USB keys, and probably facial recognition
3. Marking the password entry as insensitive if a passive method is successful, since we don't also need a password.

Code prototype updated at https://github.com/cassidyjames/agents/blob/master/src/BiometricDialog.vala

**Comment by arshubham** [comment]
@cassidyjames if a user uses passive authentication method, does he still needs to click Authenticate? A successful fingerprint scan should show us a little animation that it was successful, may be something on the lines of what is already done with system indicators like notification. Then it should automatically activate authenticate button.

**Comment by cassidyjames** [comment]
@arshubham precisely!

**Comment by ls42** [comment]
I'm really excited, that this is on the road map for eos - it would truly be a unique selling point to have an all-in-one solution baked into the base elementaryOS environment. There should be

* A user interface for registering fingerprints and also for managing existing finger prints
* Configuration for `sudo`, the permissions dialog (the one that pops up when installing packages via the app store), the login and the lock screen
* Drivers pre-installed for different hardware (these are hard to come by, apparently all newer hardware on ThinkPads have to be [reverse engineered](https://github.com/nmikhailov/Validity90))

Is there any progress or internal talk about how to progress with this?

**Comment by cassidyjames** [comment]
A couple of notes:

- USB keys are actually probably active authentication, not passive; at least the ones I've used (and most, based on messaging across most services that support them) have a physical button or contact that you have to touch to confirm authentication. Firefox and Chrome, for example, say something like "If your key has a button, press it now" and then automatically proceed once you do that. So we should do the same.
- I now have access to a Slimbook Pro X with an IR-based face unlock! It looks like they're using [Howdy](https://launchpad.net/~boltgolt/+archive/ubuntu/howdy)?
